libs/pika/execution/tests/include/algorithm_test_utils.hpp

```diff
@@ -96,6 +96,51 @@ struct error_sender
     }
 };
 
+template <typename T>
+struct const_reference_error_sender
+{
+    template <template <class...> class Tuple,
+        template <class...> class Variant>
+    using value_types = Variant<Tuple<T>>;
+
+    template <template <class...> class Variant>
+    using error_types = Variant<std::exception_ptr const&>;
+
+    static constexpr bool sends_done = false;
+
+    using completion_signatures =
+        pika::execution::experimental::completion_signatures<
+            pika::execution::experimental::set_value_t(),
+            pika::execution::experimental::set_error_t(std::exception_ptr&)>;
```

Comment:
Unsure if this will compile, but it would be the correct thing to do:
```suggestion
struct const_reference_error_sender
{
    template <template <class...> class Tuple,
        template <class...> class Variant>
    using value_types = Variant<>;

    template <template <class...> class Variant>
    using error_types = Variant<std::exception_ptr const&>;

    static constexpr bool sends_done = false;

    using completion_signatures =
        pika::execution::experimental::completion_signatures<
            pika::execution::experimental::set_error_t(std::exception_ptr&)>;
```
If the above doesn't compile do this:
```suggestion
struct const_reference_error_sender
{
    template <template <class...> class Tuple,
        template <class...> class Variant>
    using value_types = Variant<Tuple<>>;

    template <template <class...> class Variant>
    using error_types = Variant<std::exception_ptr const&>;

    static constexpr bool sends_done = false;

    using completion_signatures =
        pika::execution::experimental::completion_signatures<
            pika::execution::experimental::set_value_t(),
            pika::execution::experimental::set_error_t(std::exception_ptr&)>;
```

Since GitHub doesn't seem to highlight the changes, both variants untemplate the struct since you will never send a `T`. It's not necessary. The first variant changes the `value_types` to send nothing and removes the `set_value_t()` in `completion_signatures`. The second variant just sends no value, i.e. `Tuple<>` and keeps the `set_value_t()` in `completion_signatures`.

Reply:
So removing the `set_value` from the `completion_signatures` is working but the `Variant<>` fail on a `static_assert` checking that size of T is strictly greater than 0 so I'm going with the `Variant<Tuple<>>` solution

Reply:
Just for the record, can you put here which static assert that was?

Reply:
I'm only marking this unresolved as a reminder to write down which assertion failed. No further changes needed in this PR (for this comment).

Reply:
Also, if you have not only location of the static assertion, but the full instantiation stack that'd be helpful.

Reply:
https://gist.github.com/aurianer/8ba692582b878f57b9823a1800d2ef55

Reply:
https://github.com/pika-org/pika/blob/d723db02970b895f9f0c2f8e6b03fe1b101ec464/libs/pika/execution/include/pika/execution/algorithms/detail/single_result.hpp#L74-L75

Reply:
Thanks. I think that's a mostly a non-issue, though it could be made to work. Basically `when_all` expects that the predecessors actually send some values. If one of the predecessors only ever sends errors or stopped, `when_all` will also only ever send errors or stopped. It may useful to have this work in some generic contexts, but I think we can safely ignore it for now. The reference implementation may handle this already.

Reply:
In fact, since the CI build with the reference implementation builds without the `set_value_t()` in `completion_signatures`, it does handle this correctly.

Reply:
`completion_signatures` is only used by the reference implementation and it's able to handle no values. Our implementation does not handle it and requires an empty `Tuple<>`.